gems/aws-crt/lib/aws-crt/auth/credentials.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Aws
+  module Crt
+    module Auth
+      # CRT CredentialOptions
+      class CredentialsOptions
+        include Aws::Crt::ManagedNative
+        native_destroy Aws::Crt::Native.method(:credentials_options_release)
+
+        UINT64_MAX = 18_446_744_073_709_551_615
+
+        # @param [String] access_key_id
+        # @param [String] secret_access_key
+        # @param [String] session_token (nil)
+        # @param [Time|int] expiration (nil) - Either a Time or an int
+        #   seconds since unix epoch
+        def initialize(access_key_id, secret_access_key,
+                       session_token = nil, expiration = nil)
+          validate_access_key!(access_key_id, secret_access_key)
+
+          manage_native do
+            Aws::Crt::Native.credentials_options_new
+          end
+
+          Aws::Crt::Native.credentials_options_set_access_key_id(
+            native, access_key_id, access_key_id.length
```

Comment:
Where is `native` coming from, and why must it be passed into every function? Is there a better pattern?

Reply:
native is the managed native pointer.  We need to pass it into the functions to operate on (ie, the C functions need a pointer to the C object they are working on.  The managedNative module that gets included in these classes adds the code that links garbage collection with the destructor.  I'm open to other patterns - but when I was origionally working on the bindings this was the cleanest pattern overall that I could come up with while still being sufficiently flexible (not all of the binding functions follow the same format, so we don't want to just auto-add the native pointer as the first arg).

Reply:
Potentially a bad idea, but what about global?

Reply:
What do you mean global?  They can't be global as each of these managed objects have their own native pointer they manage.

Reply:
I was thinking some global object/array/hash that you can index into.

Reply:
What benefit would that have, other than creating a synchronization bottleneck and guaranteed lookup cost every time you access it?

Reply:
this is a parameter to the C code, we don't have objects, and creating global state is almost always a mistake because threads :)